File: frontend/lib/salesforce/data-cloud.ts
/**
 * Salesforce Data Cloud (formerly CDP / Customer Data Platform) client.
 *
 * Phase 2 of the Data 360 grounding strategy. This module calls the
 * Data Cloud Query API and Calculated Insights API to fetch wearable/EHR
 * signals that aren't available via plain SOQL against Health Cloud.
 *
 * Architecture:
 *   Phase 1 (grounding.ts)  — SOQL against Health Cloud objects
 *                              (Contact, CareProgramEnrollee, CarePlan, Case)
 *   Phase 2 (this module)   — Data Cloud Query API against Data Model Objects
 *                              (UnifiedIndividual, ssot__Observation__dlm,
 *                               ssot__CalculatedInsight__dlm, etc.)
 *
 * The two phases COMPOSE: grounding.ts calls into this module to layer
 * real calculated insights (HRV, vasomotor, sleep) on top of the SOQL-
 * derived Health Cloud insights. If this module is unconfigured or the
 * DC tenant isn't provisioned, it returns null and grounding.ts falls
 * back to the intake-baseline mocks silently.
 *
 * Env vars (all optional — absence silently disables Phase 2):
 *   SF_DC_TENANT_URL      Base URL of the Data Cloud tenant, e.g.:
 *                           https://00DHp00000L08KK.c360a.salesforce.com
 *                         In most orgs this is derivable from the SF_INSTANCE_URL
 *                         org ID; set it explicitly when the org uses a custom domain.
 *   SF_DC_API_VERSION     Data Cloud API version, default "v1". Used in path
 *                         /api/v1/query and /api/v1/insight/query.
 *   SF_CLIENT_ID          Shared with Phase 1 — the same Connected App / External
 *                         Client App is used, since DC auth is handled through the
 *                         core Salesforce token endpoint.
 *   SF_CLIENT_SECRET      Shared with Phase 1.
 *   SF_INSTANCE_URL       Used to derive SF_DC_TENANT_URL when the explicit var
 *                         is absent.
 *
 * Data Cloud Query API reference:
 *   https://developer.salesforce.com/docs/atlas.en-us.c360a_api.meta/c360a_api/
 *
 * Activation path (what to do in the Anypoint / Salesforce UI):
 *   See docs/MULESOFT_PHASE_2_DATA_CLOUD.md (generated alongside this file).
 */

import { getAccessToken, isSalesforceConfigured } from "./auth";
import type { CalculatedInsight, FederatedSource } from "../data-360";

const DC_API_VERSION = process.env.SF_DC_API_VERSION || "v1";

/**
 * Derive the Data Cloud tenant URL from SF_DC_TENANT_URL (explicit) or
 * from the 15-char org ID embedded in SF_INSTANCE_URL.
 *
 * Pattern: https://<15-char-orgId>.c360a.salesforce.com
 *
 * Returns null if we can't derive it (user needs to set SF_DC_TENANT_URL).
 */
function getDataCloudTenantUrl(): string | null {
  const explicit = process.env.SF_DC_TENANT_URL;
  if (explicit) return explicit.replace(/\/$/, "");

  const instanceUrl = process.env.SF_INSTANCE_URL;
  if (!instanceUrl) return null;

  // Extract the 18-char org ID from the subdomain, e.g.:
  //   https://trailsignup-c2d761a3b89bf2.my.salesforce.com
  // The c360a domain uses the 15-char version of the org ID from
  // /services/data/vXX.0/sobjects Organization.Id (e.g. 00DHp00000L08KK).
  // We can't derive it here without a live SOQL call, so we require the
  // explicit env var. Return null to signal misconfiguration.
  return null;
}

export function isDataCloudConfigured(): boolean {
  if (!isSalesforceConfigured()) return false;
  return getDataCloudTenantUrl() !== null;
}

/**
 * Raw shape returned by the Data Cloud Query API
 * POST /api/v1/query
 */
type DcQueryResponse = {
  data?: Array<Record<string, unknown>>;
  metadata?: Record<string, { type: string }>;
  nextBatchId?: string;
  done?: boolean;
  rowCount?: number;
};

/**
 * Raw shape returned by Data Cloud Calculated Insights API
 * GET /api/v1/insight/query?insight_api_name=<name>&filter=<expr>
 */
type DcInsightResponse = {
  data?: Array<Record<string, unknown>>;
  metadata?: Record<string, { type: string }>;
  rowCount?: number;
};

/**
 * Data Cloud uses a two-legged token flow. A normal Salesforce
 * client_credentials token (the one auth.ts mints, valid against
 * <instanceUrl>/services/data/...) is NOT directly valid against the
 * Data Cloud tenant API. It must first be exchanged at
 *   POST <instanceUrl>/services/a360/token
 * for a Data-Cloud-scoped token, which also returns the authoritative
 * tenant host. Skipping this exchange produces a 400 with an empty body
 * from the c360a gateway.
 *
 * Docs: https://developer.salesforce.com/docs/data/data-cloud-query-guide/references/data-cloud-query-api-reference/c360a-direct-api-connected-app.html
 */
type DataCloudToken = {
  accessToken: string;
  tenantUrl: string; // normalized, with scheme, no trailing slash
  expiresAtMs: number;
};

let cachedDcToken: DataCloudToken | null = null;
let inflightDcToken: Promise<DataCloudToken> | null = null;

const DC_TOKEN_SAFETY_MARGIN_MS = 60_000;
const DC_TOKEN_FALLBACK_LIFETIME_MS = 5 * 60_000;

function normalizeTenantHost(raw: string): string {
  const trimmed = raw.trim().replace(/\/+$/, "");
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return `https://${trimmed}`;
}

async function requestDataCloudToken(): Promise<DataCloudToken> {
  const { accessToken: coreToken, instanceUrl } = await getAccessToken();

  const body = new URLSearchParams({
    grant_type: "urn:salesforce:grant-type:external:cdp",
    subject_token: coreToken,
    subject_token_type: "urn:ietf:params:oauth:token-type:access_token"
  });

  const exchangeUrl = `${instanceUrl.replace(/\/+$/, "")}/services/a360/token`;
  const res = await fetch(exchangeUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Accept: "application/json"
    },
    body: body.toString(),
    cache: "no-store"
  });

  const text = await res.text();
  if (!res.ok) {
    console.warn(`[data-cloud] token exchange → ${res.status}\nbody: ${text.slice(0, 800)}`);
    throw new Error(`Data Cloud token exchange failed (${res.status})`);
  }

  const parsed = JSON.parse(text) as {
    access_token?: string;
    instance_url?: string;
    expires_in?: number;
  };

  if (!parsed.access_token || !parsed.instance_url) {
    throw new Error("Data Cloud token exchange response missing access_token or instance_url");
  }

  // Prefer the authoritative tenant host from the exchange; fall back to
  // the configured SF_DC_TENANT_URL only if the response omits it.
  const tenantUrl = normalizeTenantHost(
    parsed.instance_url || getDataCloudTenantUrl() || ""
  );
  const lifetimeMs =
    typeof parsed.expires_in === "number" && parsed.expires_in > 0
      ? parsed.expires_in * 1000
      : DC_TOKEN_FALLBACK_LIFETIME_MS;

  return {
    accessToken: parsed.access_token,
    tenantUrl,
    expiresAtMs: Date.now() + lifetimeMs - DC_TOKEN_SAFETY_MARGIN_MS
  };
}

async function getDataCloudToken(): Promise<DataCloudToken> {
  if (cachedDcToken && cachedDcToken.expiresAtMs > Date.now()) {
    return cachedDcToken;
  }
  if (!inflightDcToken) {
    inflightDcToken = requestDataCloudToken().finally(() => {
      inflightDcToken = null;
    });
  }
  cachedDcToken = await inflightDcToken;
  return cachedDcToken;
}

/** Test-only: clear the cached Data Cloud token. */
export function _resetDataCloudTokenCacheForTests(): void {
  cachedDcToken = null;
  inflightDcToken = null;
}

async function dcFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const { accessToken, tenantUrl } = await getDataCloudToken();
  const url = `${tenantUrl}${path}`;

  const res = await fetch(url, {
    ...init,
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
      Accept: "application/json",
      ...(init?.headers ?? {})
    },
    cache: "no-store"
  });

  if (!res.ok) {
    const body = await res.text();
    // Log the response body on its own line so Vercel's row-truncated log
    // viewer doesn't eat the part that explains *why* the request failed.
    console.warn(`[data-cloud] ${path} → ${res.status}\nbody: ${body.slice(0, 800)}`);
    throw new Error(`Data Cloud API ${path} failed (${res.status})`);
  }
  return res.json() as Promise<T>;
}

/**
 * Execute a Data Cloud SQL query via POST /api/v1/query.
 * Returns the rows array (empty if no results).
 */
export async function dcQuery(sql: string): Promise<Array<Record<string, unknown>>> {
  const resp = await dcFetch<DcQueryResponse>(`/api/${DC_API_VERSION}/query`, {
    method: "POST",
    body: JSON.stringify({ sql })
  });
  return resp.data ?? [];
}

/**
 * Query a named Calculated Insight via the official Data 360 endpoint:
 *   GET /api/v1/insight/calculated-insights/{ci-name}?filters=[field=value]
 *
 * The previous shape we were using — /insight/query?insight_api_name=... —
 * does not exist on this DC version and returns 400 with an empty body.
 * See https://developer.salesforce.com/docs/data/data-cloud-query-guide/references/data-cloud-query-api-reference/c360a-api-insights-ci-ci-name.html
 *
 * insightApiName  — the API Name of the CI in Data Cloud (with __cio
 *                   suffix, e.g. "Pause_HRV_RMSSD_30d__cio").
 * filterExpr      — optional [field=value] filter expression. The brackets
 *                   are part of the syntax, not template placeholders.
 *                   Example: "[unified_id__c=003Hp00003b9bdqIAA]"
 */
export async function dcInsightQuery(
  insightApiName: string,
  filterExpr?: string
): Promise<Array<Record<string, unknown>>> {
  const params = new URLSearchParams();
  if (filterExpr) params.set("filters", filterExpr);
  const queryString = params.toString();
  const path =
    `/api/${DC_API_VERSION}/insight/calculated-insights/${encodeURIComponent(insightApiName)}` +
    (queryString ? `?${queryString}` : "");
  const resp = await dcFetch<DcInsightResponse>(path);
  return resp.data ?? [];
}

// ---------------------------------------------------------------------------
// Phase 2 Calculated Insight names
//
// Data Cloud appends a "__cio" suffix to every CI's API name (Calculated
// Insight Object), the same way custom sObjects get "__c". The Developer
// Name you type in the New Insight modal becomes the bare prefix; the
// Insight Query API expects the full __cio-suffixed name.
//
// Activated on the trailsignup org 2026-06-13 (session 3).
//
// Output-column contract (load-bearing — read below) is identical across the
// two CI definitions, so this client is unchanged whether the org runs the
// mock or the real path:
//   - Mock (constants):  data-cloud/_mock_path.sql
//   - Real (aggregates the Ingestion-API-fed Pause_Wearable_Feature__dlm DMO):
//       data-cloud/Pause_HRV_RMSSD_30d.sql, Pause_Vasomotor_Burden_30d.sql,
//       Pause_Sleep_Disruption_7d.sql
// The real-data push lives in pause_ingest (examples/data_cloud_push.py).
// See docs/PHASE_2_INGESTION_API_RUNBOOK.md for the org-side wiring.
// ---------------------------------------------------------------------------

const CI_HRV_RMSSD_30D = "Pause_HRV_RMSSD_30d__cio";
const CI_VASOMOTOR_BURDEN_30D = "Pause_Vasomotor_Burden_30d__cio";
const CI_SLEEP_DISRUPTION_7D = "Pause_Sleep_Disruption_7d__cio";

/**
 * Fetch the three wearable/EHR Calculated Insights for a patient from
 * Data Cloud. These replace the intake-baseline mocks in grounding.ts.
 *
 * Returns null if Data Cloud is not configured or the tenant doesn't
 * have the CIs provisioned yet — callers fall back to the mock baseline.
 *
 * Row shapes expected (one row per insight, one column per metric).
 * All CI output columns get a "__c" suffix from the DC engine, and the
 * Dimension column we GROUP BY on is named "unified_id" (renamed from
 * the validator-incompatible "ssot__Id__c" alias):
 *   Pause_HRV_RMSSD_30d__cio       → { unified_id__c, hrv_rmssd_ms__c, z_score__c, window_days__c }
 *   Pause_Vasomotor_Burden_30d__cio → { unified_id__c, burden_score_0_100__c, flash_count_30d__c }
 *   Pause_Sleep_Disruption_7d__cio  → { unified_id__c, disruption_index_0_1__c, disrupted_nights__c }
 */
export async function getWearableInsights(unifiedPatientId: string): Promise<{
  hrv: CalculatedInsight | null;
  vasomotor: CalculatedInsight | null;
  sleep: CalculatedInsight | null;
} | null> {
  if (!isDataCloudConfigured()) return null;

  const now = new Date().toISOString();
  const src: FederatedSource = "dbdp-wearable-features";
  // Data 360 CI filter syntax is [field=value] — brackets are literal,
  // value is bare (no quotes). Filter values are case-sensitive per the
  // official docs. We pre-strip any brackets/commas from the input id to
  // avoid breaking the filter parser; Salesforce Contact IDs are
  // [a-zA-Z0-9]{15,18} so this is a safety belt, not load-bearing.
  const safeId = unifiedPatientId.replace(/[[\],=]/g, "");
  const filter = `[unified_id__c=${safeId}]`;

  try {
    const [hrvRows, vasomotorRows, sleepRows] = await Promise.all([
      dcInsightQuery(CI_HRV_RMSSD_30D, filter),
      dcInsightQuery(CI_VASOMOTOR_BURDEN_30D, filter),
      dcInsightQuery(CI_SLEEP_DISRUPTION_7D, filter)
    ]);

    const hrv: CalculatedInsight | null = hrvRows[0]
      ? {
          id: "insight.hrv-rmssd-30d",
          kind: "hrv-variability",
          name: "HRV RMSSD variability (30-day)",
          description: `RMSSD ${hrvRows[0].hrv_rmssd_ms__c} ms · z-score ${hrvRows[0].z_score__c} vs menopause cohort. Source: Oura/DBDP via Data Cloud.`,
          value: Number(hrvRows[0].z_score__c ?? 0),
          unit: "z-score",
          computedAt: now,
          sourceWindow: "30d",
          federatedFrom: [src, "jupyterhealth-fhir"]
        }
      : null;

    const vasomotor: CalculatedInsight | null = vasomotorRows[0]
      ? {
          id: "insight.vasomotor-burden-30d",
          kind: "vasomotor-burden",
          name: "Vasomotor symptom burden (30-day)",
          description: `Burden score ${vasomotorRows[0].burden_score_0_100__c}/100 · ${vasomotorRows[0].flash_count_30d__c} events in 30d. Source: wearable thermoregulation + intake via Data Cloud.`,
          value: Number(vasomotorRows[0].burden_score_0_100__c ?? 0),
          unit: "score",
          computedAt: now,
          sourceWindow: "30d",
          federatedFrom: [src, "agentforce-intake-history"]
        }
      : null;

    const sleep: CalculatedInsight | null = sleepRows[0]
      ? {
          id: "insight.sleep-disruption-7d",
          kind: "sleep-disruption",
          name: "Sleep disruption index (7-day)",
          description: `Disruption index ${sleepRows[0].disruption_index_0_1__c} · ${sleepRows[0].disrupted_nights__c} disrupted nights. Source: Oura sleep staging via Data Cloud.`,
          value: Number(sleepRows[0].disruption_index_0_1__c ?? 0),
          unit: "fraction",
          computedAt: now,
          sourceWindow: "7d",
          federatedFrom: [src, "jupyterhealth-fhir"]
        }
      : null;

    return { hrv, vasomotor, sleep };
  } catch (err) {
    // Surface to logs so silent fallbacks are debuggable in production.
    // The caller in grounding.ts still treats null as "degrade to baseline".
    console.warn(
      "[data-cloud] getWearableInsights failed; degrading to Phase 1 baseline.",
      err instanceof Error ? `${err.name}: ${err.message}` : String(err)
    );
    return null;
  }
}

File: pause_ingest/pause_ingest/data_cloud.py
"""Salesforce Data Cloud Ingestion API client (Phase 2 hardening).

This is the push side of the Phase 2 Data 360 grounding pipeline. The
frontend (``frontend/lib/salesforce/data-cloud.ts``) *reads* Calculated
Insights out of Data Cloud; this module *writes* the flattened wearable
feature rows those Insights aggregate over, straight into a Data Cloud
Ingestion API connector — no publicly-reachable JupyterHealth instance
required.

Flow (mirrors the frontend's two-legged auth exactly):

    1. client_credentials grant against
         POST <instance_url>/services/oauth2/token
       → a core Salesforce access token.
    2. CDP token exchange against
         POST <instance_url>/services/a360/token
         grant_type=urn:salesforce:grant-type:external:cdp
       → a Data-Cloud-scoped token + the authoritative tenant host.
    3. Streaming ingest:
         POST <tenant>/api/v1/ingest/sources/{connector}/{object}
         { "data": [ {record}, ... ] }

The Connected App must carry the ``cdp_ingest_api`` scope (alongside the
``cdp_query_api`` the read path uses). See
docs/PHASE_2_INGESTION_API_RUNBOOK.md for the org-side setup.

Records are plain dicts keyed by the DLO schema field names defined in
``data-cloud/Pause_Wearable_Feature.dlo-schema.json``.
"""

from __future__ import annotations

import os
from dataclasses import asdict, dataclass
from datetime import datetime
from typing import Iterable, Literal
from zoneinfo import ZoneInfo

import httpx
from dotenv import load_dotenv

ObservationType = Literal["hrv_rmssd", "sleep_session", "hot_flash", "night_sweat"]

# Default Ingestion API connector / object API names. Override via env so the
# same code targets a differently-named connector in another org.
DEFAULT_CONNECTOR = "Pause_Wearable"
DEFAULT_OBJECT = "wearable_feature"

_CDP_GRANT = "urn:salesforce:grant-type:external:cdp"


class DataCloudConfigError(RuntimeError):
    """Raised when required Data Cloud ingestion settings are missing."""


@dataclass(frozen=True)
class WearableFeatureRecord:
    """One flattened wearable-feature row destined for the Ingestion API.

    These flatten the base64-OMH FHIR Observations the ingest worker
    produces into the columnar shape Data Cloud's Calculated Insights can
    aggregate. Field names match the DLO schema JSON exactly.

    - ``record_id``: idempotency key. Re-pushing the same id upserts rather
      than duplicating, so the cohort generator is safe to re-run.
    - ``unified_id``: the Salesforce Health Cloud Contact.Id — the join key
      that resolves to ssot__Individual__dlm.ssot__Id__c / the CI
      ``unified_id__c`` dimension.
    - ``value_num``: metric value; semantics depend on ``observation_type``
      (RMSSD ms / sleep-efficiency fraction / event severity).
    """

    record_id: str
    unified_id: str
    observation_type: ObservationType
    effective_date: str  # ISO-8601
    value_num: float
    source: str

    def to_ingest_dict(self) -> dict:
        return asdict(self)


def iso_utc(dt: datetime) -> str:
    """Serialize a datetime to an ISO-8601 string Data Cloud accepts.

    Naive datetimes are assumed UTC. Data Cloud's dateTime parser wants a
    timezone offset, so we always emit one.
    """
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=ZoneInfo("UTC"))
    return dt.isoformat()


@dataclass(frozen=True)
class DataCloudIngestConfig:
    """Settings for the Data Cloud Ingestion API push.

    Reuses the same Connected App credentials as the Phase 1 SOQL path and
    the frontend read path; only the connector/object names are new.
    """

    instance_url: str
    client_id: str
    client_secret: str
    connector: str
    object_name: str

    @classmethod
    def from_env(cls, *, dotenv_path: str | None = None) -> "DataCloudIngestConfig":
        load_dotenv(dotenv_path=dotenv_path)

        def required(name: str) -> str:
            value = os.environ.get(name, "").strip()
            if not value:
                raise DataCloudConfigError(
                    f"Missing required environment variable: {name}. "
                    "See docs/PHASE_2_INGESTION_API_RUNBOOK.md."
                )
            return value

        return cls(
            instance_url=required("SF_INSTANCE_URL").rstrip("/"),
            client_id=required("SF_CLIENT_ID"),
            client_secret=required("SF_CLIENT_SECRET"),
            connector=os.environ.get("SF_DC_INGEST_CONNECTOR", DEFAULT_CONNECTOR).strip()
            or DEFAULT_CONNECTOR,
            object_name=os.environ.get("SF_DC_INGEST_OBJECT", DEFAULT_OBJECT).strip()
            or DEFAULT_OBJECT,
        )


@dataclass(frozen=True)
class DataCloudToken:
    access_token: str
    tenant_url: str  # normalized, scheme + host, no trailing slash


def _normalize_tenant_host(raw: str) -> str:
    trimmed = raw.strip().rstrip("/")
    if trimmed.startswith("http://") or trimmed.startswith("https://"):
        return trimmed
    return f"https://{trimmed}"


class DataCloudClientBase:
    """Shared two-legged Data Cloud auth for the ingest + query clients.

    Stateless except for a lazily-fetched, per-instance Data Cloud token.
    Network calls use httpx with explicit timeouts so a hung tenant can't
    wedge the caller. Subclasses add the actual ingest / query calls.
    """

    def __init__(self, config: DataCloudIngestConfig, *, timeout: float = 30.0):
        self._cfg = config
        self._timeout = timeout
        self._token: DataCloudToken | None = None

    # -- auth -----------------------------------------------------------------

    def _fetch_core_token(self, client: httpx.Client) -> tuple[str, str]:
        """client_credentials grant → (access_token, instance_url)."""
        resp = client.post(
            f"{self._cfg.instance_url}/services/oauth2/token",
            data={
                "grant_type": "client_credentials",
                "client_id": self._cfg.client_id,
                "client_secret": self._cfg.client_secret,
            },
            headers={"Accept": "application/json"},
        )
        resp.raise_for_status()
        payload = resp.json()
        token = payload.get("access_token")
        instance_url = payload.get("instance_url", self._cfg.instance_url)
        if not token:
            raise DataCloudConfigError("core token response missing access_token")
        return token, instance_url.rstrip("/")

    def _exchange_for_dc_token(
        self, client: httpx.Client, core_token: str, instance_url: str
    ) -> DataCloudToken:
        """Exchange the core token for a Data-Cloud-scoped token (a360)."""
        resp = client.post(
            f"{instance_url}/services/a360/token",
            data={
                "grant_type": _CDP_GRANT,
                "subject_token": core_token,
                "subject_token_type": "urn:ietf:params:oauth:token-type:access_token",
            },
            headers={"Accept": "application/json"},
        )
        if resp.status_code >= 400:
            # The c360a gateway returns a 400 with an often-empty body when the
            # exchange is rejected; surface what we can.
            raise DataCloudConfigError(
                f"Data Cloud token exchange failed ({resp.status_code}): "
                f"{resp.text[:400] or '<empty body>'}"
            )
        payload = resp.json()
        access = payload.get("access_token")
        tenant = payload.get("instance_url")
        if not access or not tenant:
            raise DataCloudConfigError(
                "token exchange response missing access_token or instance_url"
            )
        return DataCloudToken(
            access_token=access, tenant_url=_normalize_tenant_host(tenant)
        )

    def _get_token(self, client: httpx.Client) -> DataCloudToken:
        if self._token is None:
            core, instance_url = self._fetch_core_token(client)
            self._token = self._exchange_for_dc_token(client, core, instance_url)
        return self._token

    def check_auth(self) -> str:
        """Fetch a Data Cloud token and return the tenant URL.

        Isolates the two-legged auth so the preflight can tell an auth problem
        (bad creds / Data Cloud not enabled / missing CDP grant) apart from a
        data problem (DMO or CI not created yet). Raises on any auth failure.
        """
        with httpx.Client(timeout=self._timeout) as client:
            return self._get_token(client).tenant_url


class DataCloudIngestClient(DataCloudClientBase):
    """Thin Data Cloud Ingestion API client (the write path).

    Construct with a config; call :meth:`ingest`. Needs the ``cdp_ingest_api``
    scope on the Connected App.
    """

    # -- ingest ---------------------------------------------------------------

    def ingest(self, records: Iterable[WearableFeatureRecord]) -> dict:
        """Push records to the Ingestion API. Returns the parsed response.

        Records are sent in a single ``{"data": [...]}`` batch. The Streaming
        Ingestion API accepts up to 200 records / request; callers with more
        should chunk (see :func:`chunked`).
        """
        batch = [r.to_ingest_dict() for r in records]
        if not batch:
            return {"accepted": 0, "skipped": "empty batch"}

        with httpx.Client(timeout=self._timeout) as client:
            token = self._get_token(client)
            path = (
                f"{token.tenant_url}/api/v1/ingest/sources/"
                f"{self._cfg.connector}/{self._cfg.object_name}"
            )
            resp = client.post(
                path,
                json={"data": batch},
                headers={
                    "Authorization": f"Bearer {token.access_token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
            if resp.status_code >= 400:
                raise DataCloudConfigError(
                    f"ingest POST failed ({resp.status_code}): {resp.text[:600]}"
                )
            # The Ingestion API returns 202 Accepted with an empty/short body.
            try:
                body = resp.json()
            except ValueError:
                body = {}
            return {"accepted": len(batch), "status_code": resp.status_code, **body}


class DataCloudQueryClient(DataCloudClientBase):
    """Read Calculated Insights back out of Data Cloud (the verify path).

    The read-back counterpart to :class:`DataCloudIngestClient`: after the push
    + CI refresh, this queries each activated Calculated Insight exactly the
    way the frontend does (``frontend/lib/salesforce/data-cloud.ts``) so a
    maintainer can prove the CIs return real per-patient values instead of the
    ``MAX(constant)`` mock. Needs the ``cdp_query_api`` scope on the Connected
    App (the read path already relies on it).
    """

    def query_calculated_insight(
        self, insight_api_name: str, filter_expr: str | None = None
    ) -> list[dict]:
        """GET one Calculated Insight's rows, optionally filtered.

        Mirrors the frontend read exactly:
            GET {tenant}/api/v1/insight/calculated-insights/{name}?filters=[field=value]

        ``filter_expr`` is the bare bracketed expression the Data 360 Insight
        API expects, e.g. ``[unified_id__c=003Hp00003b9bdqIAA]`` — the brackets
        are literal syntax, and httpx percent-encodes them the same way the
        frontend's URLSearchParams does. Returns the ``data`` array (empty when
        the CI has no row for the filter).
        """
        params = {"filters": filter_expr} if filter_expr else None
        with httpx.Client(timeout=self._timeout) as client:
            token = self._get_token(client)
            path = (
                f"{token.tenant_url}/api/v1/insight/calculated-insights/"
                f"{insight_api_name}"
            )
            resp = client.get(
                path,
                params=params,
                headers={
                    "Authorization": f"Bearer {token.access_token}",
                    "Accept": "application/json",
                },
            )
            if resp.status_code >= 400:
                raise DataCloudConfigError(
                    f"insight query for {insight_api_name} failed "
                    f"({resp.status_code}): {resp.text[:600]}"
                )
            try:
                body = resp.json()
            except ValueError:
                return []
            data = body.get("data")
            return data if isinstance(data, list) else []

    def query(self, sql: str) -> list[dict]:
        """Run a Data Cloud SQL query via ``POST /api/v1/query``.

        Mirrors the frontend's ``dcQuery`` (data-cloud.ts). Used by the
        preflight to probe whether the ``Pause_Wearable_Feature__dlm`` DMO
        exists and holds rows; raises :class:`DataCloudConfigError` on an HTTP
        error (e.g. the DMO not existing yet).
        """
        with httpx.Client(timeout=self._timeout) as client:
            token = self._get_token(client)
            resp = client.post(
                f"{token.tenant_url}/api/v1/query",
                json={"sql": sql},
                headers={
                    "Authorization": f"Bearer {token.access_token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
            if resp.status_code >= 400:
                raise DataCloudConfigError(
                    f"query failed ({resp.status_code}): {resp.text[:600]}"
                )
            try:
                body = resp.json()
            except ValueError:
                return []
            data = body.get("data")
            return data if isinstance(data, list) else []


def chunked(records: list[WearableFeatureRecord], size: int = 200) -> Iterable[list]:
    """Yield successive ``size``-length chunks (Ingestion API caps at 200)."""
    for i in range(0, len(records), size):
        yield records[i : i + size]


def build_ingest_payload(records: Iterable[WearableFeatureRecord]) -> dict:
    """Build the ``{"data": [...]}`` payload without sending it.

    Used by the dry-run path of the example script and by tests, so the
    record→payload shaping is verifiable without a live tenant.
    """
    return {"data": [r.to_ingest_dict() for r in records]}
